godot-core/src/obj/raw.rs

```diff
@@ -42,30 +42,34 @@ impl<T: GodotClass> RawGd<T> {
         }
     }
 
-    /// Initializes this `RawGd<T>` from the object pointer as a **weak ref**, meaning it does not
-    /// initialize/increment the reference counter.
-    ///
     /// # Safety
     ///
     /// `obj` must be a valid object pointer or a null pointer.
-    pub(super) unsafe fn from_obj_sys_weak(obj: sys::GDExtensionObjectPtr) -> Self {
-        let rtti = if obj.is_null() {
+    unsafe fn get_rtti_of(ptr: sys::GDExtensionObjectPtr) -> Option<ObjectRtti> {
+        if ptr.is_null() {
             None
         } else {
-            let raw_id =
-                unsafe { interface_fn!(object_get_instance_id)(obj as sys::GDExtensionObjectPtr) };
+            let raw_id = unsafe { interface_fn!(object_get_instance_id)(ptr) };
 
             let instance_id = InstanceId::try_from_u64(raw_id)
                 .expect("constructed RawGd weak pointer with instance ID 0");
 
             // TODO(bromeon): this should query dynamic type of object, which can be different from T (upcast, FromGodot, etc).
             // See comment in ObjectRtti.
             Some(ObjectRtti::of::<T>(instance_id))
-        };
+        }
+    }
 
+    /// Initializes this `RawGd<T>` from the object pointer as a **weak ref**, meaning it does not
+    /// initialize/increment the reference counter.
+    ///
+    /// # Safety
+    ///
+    /// `obj` must be a valid object pointer or a null pointer.
+    pub(super) unsafe fn from_obj_sys_weak(obj: sys::GDExtensionObjectPtr) -> Self {
         Self {
-            obj: obj as *mut T,
-            cached_rtti: rtti,
+            obj: obj.cast::<T>(),
+            cached_rtti: Self::get_rtti_of(obj),
```

Comment:
Any reason why this is extracted? Since it's only used once, the code is easier to read without indirection, and it's immediately clear this isn't needed somewhere else. It's not like the function was hundreds of lines long🙂

Reply:
oh i can unextract it. forgot about that. i tried a different solution in which i needed to initialize the rtti in both `new_with_uninit` and `new_with_init`, but since this just reimplements `new_with_init` as calling `new_with_init` that's unnecessary.